### PR TITLE
Skip adding builtins if they will be removed

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
When loading from a snapshot, a validator adds and removes builtins and precompiles according to which runtime features are activated. When loading from a snapshot, all activated builtins are added before processing any removals. So after a builtin has been removed, the validator will still try to re-add it whenever it starts up from a snapshot. Since the builtin doesn't exist anymore on-chain, it will hard exit because it thinks it booted from an incompatible snapshot.

#### Summary of Changes
- Add additional types for feature activated transitions to the list of active builtin programs
- Ensure that any removed builtin references an existing builtin that is retained if the removal feature hasn't been activated yet

Fixes https://github.com/solana-labs/solana/issues/23231

### CODE REVIEW

File: ci/test-checks.sh [hunk 1]
```diff
@@ -44,6 +44,8 @@ echo --- build environment
 export RUST_BACKTRACE=1
 export RUSTFLAGS="-D warnings -A incomplete_features"
 
+_ "$cargo" stable clean
```
Comment (ryoqun) [inline ci/test-checks.sh:47]: is this intended to be included in this pr? i think this increase the build time sometimes needlessly by removing the build cache too much?

Comment (jstarry) [reply ci/test-checks.sh:47]: It was my only way to get past a compiler issue on CI. I will remove before merging. I mentioned it in #devops on Discord as well

**Comment by jstarry** [comment]
Oops I still need to fix the Abi changes. But the logic of the PR should still be reviewable in its current form

File: runtime/src/builtins.rs [hunk 2]
```diff
@@ -51,11 +51,61 @@ macro_rules! with_program_logging {
...
+    Add {
+        builtin: Builtin,
+        feature_id: Pubkey,
+    },
+    /// Remove a builtin program if a feature is activated or
+    /// retain a previously added builtin.
+    RemoveOrRetain {
+        previous_builtin: Builtin,
+        removal_feature_id: Pubkey,
+    },
```
[the remark this answers is not in the document]
Comment (ryoqun) [reply runtime/src/builtins.rs:68]: also, I'm omitting `Replace` as it's really useless now. That's because builtins now has access to feature_set. `Replace` was relic before the widespread use of feature activation mechanism...

[the remark this answers is not in the document]
Comment (ryoqun) [reply runtime/src/builtins.rs:68]: lastly, i wonder when `Remove` would be needed in the (distant) future. so, I'm just saying leaving a comment to work as a pointer to this pr.

[the remark this answers is not in the document]
Comment (jstarry) [reply runtime/src/builtins.rs:68]: Yes, that's exactly why I removed Replace. Relic of the past and not useful anymore.

[the remark this answers is not in the document]
Comment (jstarry) [reply runtime/src/builtins.rs:68]: Great yeah I will add a comment, I agree it probably won't be needed in the future

File: runtime/src/bank/builtin_programs.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#[cfg(test)]
+mod tests {
+    use {
+        crate::bank::*,
+        solana_sdk::{feature_set::FeatureSet, genesis_config::create_genesis_config},
+    };
+
+    #[test]
+    fn test_startup_from_snapshot_after_precompile_transition() {
```
Comment (ryoqun) [inline runtime/src/bank/builtin_programs.rs:9]: thanks for writing this. yeah, feature activation needs more test coverage...

File: runtime/src/builtins.rs [hunk 4]
```diff
@@ -150,73 +190,61 @@ fn dummy_process_instruction(
...
                 "compute_budget_program",
                 solana_sdk::compute_budget::id(),
                 solana_compute_budget_program::process_instruction,
             ),
-            feature_set::add_compute_budget_program::id(),
-            ActivationType::NewProgram,
-        ),
+            feature_id: feature_set::add_compute_budget_program::id(),
+        },
         // TODO when feature `prevent_calling_precompiles_as_programs` is
```
Comment (jackcmay) [inline runtime/src/builtins.rs:204]: These comments can be removed

**Comment by jackcmay** [comment]
Looks good @jstarry thanks!

On main and devnet all existing builtin and precompile accounts are owned by the native loader.  I added a change to this pr to keep that consistency.

Also added a few nits

File: runtime/src/builtins.rs [hunk 5]
```diff
@@ -51,63 +51,6 @@ macro_rules! with_program_logging {
     };
 }
 
-/// State transition enum used for adding and removing builtin programs through
```
Comment (t-nelson) [inline runtime/src/builtins.rs]: This block and the next are purely cosmetic, right?  Hard to follow major cosmetics in a functional change set

**Review by t-nelson — APPROVED** [review]
lgtm. thanks!  third times the charm

**Comment by jstarry** [comment]
Cool, I would prefer the owner fix to be separate but looks safe

**Comment by t-nelson** [comment]
let's go?

File: runtime/src/builtins.rs [hunk 6]
```diff
@@ -101,8 +94,63 @@ pub struct Builtins {
...
+            Self::RemoveOrRetain {
+                previous_builtin,
+                removal_feature_id,
+            } => {
+                if should_apply_action_for_feature(removal_feature_id) {
+                    Some(BuiltinAction::Remove(previous_builtin.id))
+                } else {
+                    // Retaining is no different from adding a new builtin.
+                    Some(BuiltinAction::Add(previous_builtin.clone()))
+                }
```
Comment (jstarry) [inline runtime/src/builtins.rs:150]: I believe this branch introduced a new bug for epoch boundaries. On epoch boundaries, the `should_apply_action_for_feature(removal_feature_id)` condition will be false if the removal feature id wasn't newly activated. This causes the builtin to be re-added even if it was previously removed.
